src/runtime-rs/crates/runtimes/virt_container/src/container_manager/manager.rs

```diff
@@ -0,0 +1,274 @@
+                let containers = self.containers.read().await;
+                let c = containers
+                    .get(container_id)
+                    .ok_or_else(|| Error::ContainerNotFound(container_id.to_string()))?;
+                c.delete_exec_process(process)
+                    .await
+                    .context("delete process")?;
+                c.state_process(process).await.context("state process")
+            }
+        }
+    }
+
+    async fn exec_process(&self, req: ExecProcessRequest) -> Result<()> {
+        if req.spec_type_url.is_empty() {
+            return Err(anyhow!("invalid type url"));
+        }
+        let oci_process: OCIProcess =
+            serde_json::from_slice(&req.spec_value).context("serde from slice")?;
+
+        let containers = self.containers.read().await;
+        let container_id = &req.process.container_id.container_id;
+        let c = containers
+            .get(container_id)
+            .ok_or_else(|| Error::ContainerNotFound(container_id.clone()))?;
+        c.exec_process(
+            &req.process,
+            req.stdin,
+            req.stdout,
+            req.stderr,
+            req.terminal,
+            oci_process,
+        )
+        .await
+        .context("exec")?;
+        Ok(())
+    }
+
+    async fn kill_process(&self, req: &KillRequest) -> Result<()> {
+        let containers = self.containers.read().await;
+        let container_id = &req.process.container_id.container_id;
+        let c = containers
+            .get(container_id)
+            .ok_or_else(|| Error::ContainerNotFound(container_id.clone()))?;
+        c.kill_process(&req.process, req.signal, req.all)
+            .await
+            .map_err(|err| {
+                warn!(
+                    sl!(),
+                    "failed to signal process {:?} {:?}", &req.process, err
+                );
+                err
+            })
+            .ok();
+        Ok(())
+    }
+
+    async fn wait_process(&self, process: &ContainerProcess) -> Result<ProcessExitStatus> {
+        let logger = logger_with_process(process);
+
+        let containers = self.containers.read().await;
+        let container_id = &process.container_id.container_id;
+        let c = containers
+            .get(container_id)
+            .ok_or_else(|| Error::ContainerNotFound(container_id.clone()))?;
+        let (watcher, status) = c.wait_process(process).await.context("wait")?;
+        drop(containers);
+
+        match watcher {
+            Some(mut watcher) => {
+                info!(logger, "begin wait exit");
+                while watcher.changed().await.is_ok() {}
+                info!(logger, "end wait exited");
+            }
+            None => {
+                warn!(logger, "failed to find watcher for wait process");
+            }
+        }
+
+        let status = status.read().await;
+
+        info!(logger, "wait process exit status {:?}", status);
+
+        // stop process
```

Comment:
Question: why call `stop_process` here?

Reply:
Thanks for your review, this operation tries to clean up the container/process when the container/process exit.